Enable clang-tidy for headers

rschu1ze opened the pull request:
clang-tidy now also checks code in header files. Because the analyzer
finds tons of issues, activate the check only for directory "base/" (see
file ".clang-tidy"). All other directories, in particular "src/" are
left to future work (--> #40741)

While many findings were fixed, some were not (and suppressed instead).
Reasons for this include: a) the file is 1:1 copypaste of a 3rd-party
lib (e.g. pcg_extras.h) and fixing stuff would make upgrades/fixes more
difficult b) a fix would have broken lots of using code

See also https://github.com/ClickHouse/ClickHouse/issues/6044#issuecomment-603958394

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

rschu1ze committed:
Enable clang-tidy for headers

clang-tidy now also checks code in header files. Because the analyzer
finds tons of issues, activate the check only for directory "base/" (see
file ".clang-tidy"). All other directories, in particular "src/" are
left to future work.

While many findings were fixed, some were not (and suppressed instead).
Reasons for this include: a) the file is 1:1 copypaste of a 3rd-party
lib (e.g. pcg_extras.h) and fixing stuff would make upgrades/fixes more
difficult b) a fix would have broken lots of using code

antonio2368 commented:
>  it is mostly copy paste of external
libraries where we should keep changes few and small

I would agree with any changes in logic, but is this true for clang-tidy changes? They should be just a better way of writing the same thing so why not fix it in those files also?
How did you decide which files to exclude from checks?

rschu1ze commented:
I changed it so that more stuff is fixed. In fact when looking closer, only "pcg_extras.h" is 1:1 copypaste of a 3rd party lib. I really rather don't touch this file as that makes fixes/upgrades more difficult. I fixed a lot of stuff in the other files but even there were places where a proper fix would have broken too much using code.

rschu1ze committed:
fix: typo

rschu1ze committed:
Merge branch 'master' into clang-tidy-for-headers

antonio2368 commented on base/base/itoa.h:
```diff
@@ -352,7 +352,7 @@ static inline char * writeUIntText(T x, char * p)
 
     int len = digits10(x);
-    auto pp = p + len;
+    auto *pp = p + len;
```
```suggestion
    auto * pp = p + len;
```

rschu1ze committed:
Update base/base/itoa.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>
